scripts/autogen-version

```diff
@@ -1,166 +1,147 @@
-        ref=HEAD
-        describe_arg="--dirty"
-
-        test -n "$1" && test "$1" != "HEAD" && ref="$1" && describe_arg="$1"
-
-        hash=$(git rev-parse --short "${ref}")
-        branch=$(git symbolic-ref --short "${ref}" 2>/dev/null || echo "${hash}")
-        git_version=$(git describe --always --tags --match "v*" "${describe_arg}" | sed 's/^v//' | sed 's/-\([0-9]*\)-g[0-9a-z]*/.\1/g'
-    )
-    elif [ -f ./VERSION ]; then
-        # Read information from VERSION file which contains values for `branch`, `hash`, and `git_version`.
-        . ./VERSION
-    else
-        >&2 echo "Cannot extract version information: neither git repository nor VERSION file present"
-        exit 1
-    fi
-}
+import argparse
+import os
+import re
+import shutil
+import subprocess
+import tempfile
+
 
-get_version "$@"
+def get_version(ref):
+    git_tree = subprocess.run(
+        ["git", "rev-parse", "--git-dir"],
+        stdout=subprocess.DEVNULL,
+        stderr=subprocess.DEVNULL,
+    )
+    if git_tree.returncode == 0:
+        hash = subprocess.check_output(
+            ["git", "rev-parse", "--short", ref], universal_newlines=True
+        ).strip()
+
+        branch = hash
+        try:
+            branch = subprocess.check_output(
+                ["git", "symbolic-ref", "--short", ref],
+                universal_newlines=True,
+                stderr=subprocess.DEVNULL,
+            ).strip()
+        except:
+            pass
+
+        describe_arg = ref if ref != "HEAD" else "--dirty"
+        raw_git_version = subprocess.check_output(
+            ["git", "describe", "--always", "--tags", "--match", "v*", describe_arg],
+            universal_newlines=True,
+        )
+        git_version = re.sub(
+            r"-([0-9]*)-g[0-9a-z]*", r".\1", re.sub(r"^v", "", raw_git_version)
+        ).strip()
+
+        return hash, branch, git_version
+    elif os.path.exists("VERSION"):
+        with open("VERSION", "r") as info:
+            hash = info.readline().split("=")[1].strip()
+            branch = info.readline().split("=")[1].strip()
+            git_version = info.readline().split("=")[1].strip()
+            return hash, branch, git_version
+    else:
+        print(
+            "Cannot extract version information: neither git repository nor VERSION file present"
+        )
+        exit(1)
+
+
+parser = argparse.ArgumentParser(
+    description="Generates a readable representation of the current version number"
+)
+output_mode_args = parser.add_mutually_exclusive_group()
+output_mode_args.add_argument("--short", action="store_true")
+output_mode_args.add_argument("--cmake", action="store_true")
+parser.add_argument("--header", metavar="file", nargs="?", const="version.h")
+parser.add_argument("--store", metavar="file", nargs="?", const="VERSION")
+parser.add_argument("--git-root", metavar="directory", nargs=1)
+parser.add_argument("--commit", metavar="COMMIT_HASH", default="HEAD")
```

Comment:
While we are at it we could add some meaningful documentation on the semantics of these flags, e.g.,
```suggestion
output_mode_args.add_argument(
    "--short", action="store_true", help="output a short version string")
output_mode_args.add_argument(
    "--cmake", action="store_true", help="output version string for use with CMake")
parser.add_argument("--header", metavar="file", nargs="?",
                    const="version.h", help="generate a version.h file for compilation")
parser.add_argument("--store", metavar="file", nargs="?", const="VERSION",
                    help="generate VERSION file persisting Git information")
parser.add_argument("--git-root", metavar="directory", nargs=1,
                    help="path to directory containing .git folder")
parser.add_argument("--commit", metavar="COMMIT_HASH", default="HEAD",
                    help="commit to generate version information for")
```